packages/babel-parser/src/plugins/typescript/index.js

```diff
@@ -2779,6 +2779,20 @@ export default (superClass: Class<Parser>): Class<Parser> =>
       return super.parseMaybeDecoratorArguments(expr);
     }
 
+    checkCommaAfterRest(close) {
+      if (this.match(tt.comma)) {
+        if (this.lookaheadCharCode() === close) {
+          if (!this.state.isDeclareContext) {
```

Comment:
Is this condition true in `d.ts` files?

Reply:
It is true if there is the `declare` keyword before the function declaration.

Reply:
In that case this PR doesn't fully fix #13100

Reply:
We don't have a separate parsing goal for `.d.ts` files I think, since any code that is valid in `.d.ts` files was valid in normal `.ts` files.

Reply:
(Q: Do you know if I can use the `.d.ts` parse goal on the TS playground?)

Reply:
The problem is that
`function foo(...args: any[], )` (without the body) is invalid in `.ts` files even if it is only a type definition.

Reply:
> any code that is valid in .d.ts files was valid in normal .ts

```ts
class Foo {
    foo(): any;
}
```
is invalid in `.ts` unless it's nested in `declare namespace`.

As for `function foo(...args: any[], )` without the body, it can be used for two different things:

1) for [overloads](https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads)
2) for type-only declarations (inside `declare namespace`, `declare global`, `declare module "name" { ... }`, and probably something else)

In the first case, the rest parameter trailing comma isn't allowed. In the latter, it's allowed as it's an "ambient context".

Reply:
> Do you know if I can use the .d.ts parse goal on the TS playground

I believe if you wrap you code e.g. in `declare module 'foo' { ... }`, it'll be the same.

Reply:
I think we can start by fixing the "easy" case (which is what this PR does), but we'll eventually need to add a `{ dts: boolean }` option to the typescript parser plugin.

Reply:
From Prettier's perspective, making the error recoverable would suffice.

Reply:
I've created a new issue so we don't forget #13108